### PR TITLE
feat(server): add builtin tools system with tasks tool

### PR DESCRIPTION
**Pull request by BobbyRadford**
## Summary

Add a builtin tools framework for the chat agent. Every runtime model configuration now gets a set of built-in tools automatically. Users can disable specific tools by name via an `exclude` list but cannot modify tool schemas.

The first builtin tool is `tasks` - a task management tool similar to Claude Code's task tracking system.

## Key Features

- **TasksService**: File-based persistence at `~/.llamafarm/projects/{ns}/{proj}/tasks/{session_id}/`
  - Auto-incrementing IDs (1.json, 2.json, ...)
  - File locking via `filelock` for concurrent subagent access
  - Bidirectional blocking relationships (blockedBy/blocks)
  - Cycle detection to prevent circular dependencies

- **Builtin Tools Registry**: Hardcoded tool definitions with config-based enable/exclude
  - New `builtin_tools` config field on Model schema
  - `enabled: bool` master switch (default: true)
  - `exclude: list[str]` to disable specific tools by name

- **Tasks Tool**: Full CRUD operations
  - Operations: create, update, list, get (delete via update with status="deleted")
  - Status lifecycle: pending → in_progress → completed
  - Automatic unblocking when dependencies complete

- **Agent Integration**: Refactored tool execution architecture
  - `_execute_tool()` unified dispatcher
  - `_execute_mcp_tool()` for MCP tools
  - `_execute_builtin_tool()` for builtin tools
  - Updated `_can_execute_tool_call()` to check both tool types

## Files Changed

| File | Changes |
|------|---------|
| `config/schema.yaml` | Added `builtin_tools` to Model schema |
| `server/services/tasks_service.py` | New - Core task CRUD with file locking |
| `server/tools/builtin/registry.py` | New - `BUILTIN_TOOLS` and filtering |
| `server/tools/builtin/factory.py` | New - Context-injecting factory |
| `server/tools/builtin/tasks_tool.py` | New - Tasks tool implementation |
| `server/agents/chat_orchestrator.py` | Integrated builtin tools |
| `server/agents/base/types.py` | Made `from_mcp_tool()` defensive |

## Test Plan

120 new tests added across 3 test files:
- `test_tasks_service.py` - 59 tests (CRUD, locking, cycles, stress)
- `test_builtin_tools.py` - 40 tests (registry, factory, tool, schema)
- `test_agent_builtin_tools.py` - 21 tests (loading, dispatch, merging)

```bash
# Run all new tests
cd server && uv run pytest tests/test_tasks_service.py tests/test_builtin_tools.py tests/test_agent_builtin_tools.py -v

# Run full test suite (589 existing + 120 new = 709 total)
cd server && uv run pytest tests/ --ignore=tests/test_voice*.py
```

## Example Usage

```yaml
# In llamafarm.yaml - disable tasks tool
runtime:
  models:
    - name: default
      provider: universal
      model: unsloth/Qwen3-1.7B-GGUF:Q4_K_M
      builtin_tools:
        exclude: ["tasks"]
```

```
# In chat - use tasks tool
> Create a task to implement user authentication
> List my tasks
> Mark task 1 as in progress
> Create a task blocked by task 1
```

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- **Builtin Tools Framework**: Introduced a comprehensive system for built-in tools that are automatically available to chat agents, with configuration-based enable/disable controls

- **TasksService**: New file-based task management service with auto-incrementing IDs, file locking for concurrent access, bidirectional dependency tracking, and cycle detection

- **Tasks Tool**: Full CRUD implementation (create, update, list, get) with status lifecycle (pending → in_progress → completed) and automatic unblocking of dependent tasks

- **Tool Registry & Factory**: New registry system for defining builtin tools and factory for creating instances with injected context (`_project_dir`, `_session_id`)

- **Agent Integration**: Refactored `ChatOrchestratorAgent` to load and execute builtin tools alongside MCP tools with unified dispatcher (`_execute_tool()`)

- **Configuration**: Added `builtin_tools` config field to Model schema with `enabled` flag and `exclude` list for selective tool disabling

- **Comprehensive Testing**: Added 120 new tests across 3 test files covering CRUD operations, concurrency, cycle detection, tool registry, factory, and agent integration

- **Defensive Improvements**: Enhanced `from_mcp_tool()` with safer attribute access and null-coalescing for missing tool metadata

___

### Diagram Walkthrough

```mermaid
flowchart LR
  Config["Config Schema<br/>builtin_tools"]
  Registry["Builtin Tools<br/>Registry"]
  Factory["Builtin Tool<br/>Factory"]
  TasksService["TasksService<br/>File-based Persistence"]
  TasksTool["TasksTool<br/>CRUD Operations"]
  Agent["ChatOrchestratorAgent<br/>Tool Execution"]
  
  Config -- "defines enabled/exclude" --> Registry
  Registry -- "filters tools" --> Factory
  Factory -- "creates instances<br/>with context" --> TasksTool
  TasksTool -- "uses" --> TasksService
  Factory -- "provides tools to" --> Agent
  Agent -- "executes via<br/>_execute_tool()" --> TasksTool
```

<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Tests</strong></td><td><details><summary>3 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>test_tasks_service.py</strong><dd><code>Comprehensive TDD test suite for TasksService with 59 tests</code></dd></summary>
<hr>

server/tests/test_tasks_service.py

<ul><li>Added 1754 lines of comprehensive TDD tests for <code>TasksService</code> covering <br>CRUD operations, dependency management, cycle detection, and <br>concurrency handling<br> <li> Organized into 9 test classes: <code>TestTaskCreation</code>, <code>TestTaskRetrieval</code>, <br><code>TestTaskUpdate</code>, <code>TestTaskDeletion</code>, <code>TestCycleDetection</code>, <code>TestConcurrency</code>, <br><code>TestEdgeCases</code>, <code>TestConcurrencyStress</code>, and <code>TestComplexDependencyGraphs</code><br> <li> Tests cover task persistence, file locking, bidirectional blocking <br>relationships, auto-incrementing IDs, and stress testing with 50 <br>concurrent operations<br> <li> Includes edge cases like Unicode handling, very long descriptions, <br>duplicate dependencies, and lock timeout scenarios</ul>

</details>

  </td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/738/files#diff-bf81d854aa91aa96c06ab4c6bade99926cecfece1badbac832a89918315bf9cb">+1754/-0</a></td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>test_builtin_tools.py</strong><dd><code>Comprehensive TDD test suite for builtin tools with 40 tests</code></dd></summary>
<hr>

server/tests/test_builtin_tools.py

<ul><li>Added 889 lines of comprehensive TDD tests for builtin tools <br>infrastructure covering registry, factory, and tool execution<br> <li> Organized into 7 test classes: <code>TestGetEnabledBuiltinTools</code>, <br><code>TestBuiltinToolFactory</code>, <code>TestTasksToolSchema</code>, <code>TestTasksToolOperations</code>, <br><code>TestTasksToolErrorHandling</code>, <code>TestTasksToolWithDependencies</code>, and <br><code>TestBuiltinToolsIntegration</code><br> <li> Tests verify tool filtering based on config, context injection into <br>tool instances, and full CRUD operations via the tasks tool<br> <li> Includes schema validation, error handling for missing tasks, and <br>integration tests combining registry and factory</ul>

</details>

  </td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/738/files#diff-88dfd094f0bd8288b4371a6a685dd6e63aef47ac168db0eb278a01348a72d307">+889/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>test_agent_builtin_tools.py</strong><dd><code>Comprehensive TDD test suite for agent builtin tools with 21 tests</code></dd></summary>
<hr>

server/tests/test_agent_builtin_tools.py

<ul><li>Added 802 lines of comprehensive TDD tests for <code>ChatOrchestratorAgent</code> <br>integration with builtin tools<br> <li> Organized into 5 test classes: <code>TestBuiltinToolsLoading</code>, <br><code>TestToolExecutionDispatch</code>, <code>TestToolMerging</code>, <code>TestToolDetection</code>, and <br><code>TestBuiltinToolsAgentIntegration</code><br> <li> Tests verify agent loads builtin tools via <code>_load_builtin_tools()</code>, <br>dispatches to <code>_execute_builtin_tool()</code>, merges tools from multiple <br>sources, and detects both MCP and builtin tools<br> <li> Includes full orchestration loop testing and factory integration</ul>

</details>

  </td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/738/files#diff-7bca5d1366b08ec63f1929cb499e74602895d93a67158904e4c7dba628502e3f">+802/-0</a>&nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Enhancement</strong></td><td><details><summary>6 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>tasks_tool.py</strong><dd><code>New TasksTool implementation with full CRUD operations</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

server/tools/builtin/tasks_tool.py

<ul><li>New file implementing <code>TasksTool</code> class extending <code>BaseTool</code> for task <br>management operations<br> <li> Defines <code>TasksToolInput</code> and <code>TasksToolOutput</code> schemas with support for <br>create, update, list, and get operations<br> <li> Implements async <code>arun()</code> method dispatching to operation handlers <br>(<code>_create_task</code>, <code>_update_task</code>, <code>_list_tasks</code>, <code>_get_task</code>)<br> <li> Includes context injection via class-level <code>_project_dir</code> and <br><code>_session_id</code> attributes for accessing <code>TasksService</code></ul>

</details>

  </td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/738/files#diff-898f1e22ca87ade2e838e98aecb07b2d0bb732faf4346da5b1de5ff77259daa5">+239/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>tasks_service.py</strong><dd><code>Task management service with persistence and dependency tracking</code></dd></summary>
<hr>

server/services/tasks_service.py

<ul><li>New file implementing <code>TasksService</code> for CRUD operations on tasks with <br>file-based persistence<br> <li> Provides auto-incrementing task IDs, file locking for concurrency, and <br>bidirectional dependency tracking<br> <li> Includes cycle detection to prevent circular task dependencies<br> <li> Supports task status lifecycle (pending → in_progress → completed) <br>with automatic unblocking</ul>

</details>

  </td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/738/files#diff-cc8ce5e00ab185d971b749661f3a294347cfeb7c7e2c7cc9df54cef9054fcf1a">+482/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>registry.py</strong><dd><code>Builtin tools registry and configuration filtering</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

server/tools/builtin/registry.py

<ul><li>New file defining <code>BUILTIN_TOOLS</code> registry with task management tool <br>definition<br> <li> Implements <code>get_enabled_builtin_tools()</code> function to filter tools based <br>on model configuration<br> <li> Provides schema for tasks tool with operations (create, update, list, <br>get) and parameters</ul>

</details>

  </td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/738/files#diff-02a8097ebc78287f487083f1a3e114facbc4fdfd5838873690d255863f93dc9d">+97/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>factory.py</strong><dd><code>Factory for creating builtin tools with context injection</code></dd></summary>
<hr>

server/tools/builtin/factory.py

<ul><li>New file implementing <code>BuiltinToolFactory</code> for creating tool instances <br>with injected context<br> <li> Provides <code>create_tasks_tool()</code> and <code>create_all_tools()</code> methods<br> <li> Injects <code>_project_dir</code> and <code>_session_id</code> into tool classes for stateful <br>operations</ul>

</details>

  </td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/738/files#diff-a58e48ce82f05a254e1e1a82255cf971593fd8ff2fed141c45ec0759ead3059f">+62/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>__init__.py</strong><dd><code>Builtin tools module initialization and exports</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

server/tools/builtin/__init__.py

<ul><li>New file providing module exports for builtin tools infrastructure<br> <li> Exports <code>BUILTIN_TOOLS</code>, <code>get_enabled_builtin_tools</code>, <code>BuiltinToolFactory</code>, <br>and <code>TasksTool</code></ul>

</details>

  </td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/738/files#diff-3c2acb59f9b1edd887e4bf08bb59efd2020098a7c384aff4e4f8900df6837c90">+23/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>chat_orchestrator.py</strong><dd><code>Integrate builtin tools into chat orchestrator agent</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

server/agents/chat_orchestrator.py

<ul><li>Added imports for <code>BuiltinToolFactory</code> and <code>get_enabled_builtin_tools</code><br> <li> Added <code>_builtin_tools</code> list to store loaded builtin tool classes<br> <li> Implemented <code>_load_builtin_tools()</code> method to initialize builtin tools <br>with project context<br> <li> Refactored tool execution to unified <code>_execute_tool()</code> dispatcher that <br>routes to MCP or builtin executors<br> <li> Implemented <code>_execute_builtin_tool()</code> for executing builtin tool <br>instances<br> <li> Updated <code>_can_execute_tool_call()</code> to check both MCP and builtin tools<br> <li> Modified <code>run_async()</code> and <code>run_async_stream()</code> to merge builtin tool <br>definitions with MCP tools<br> <li> Updated <code>setup_tools()</code> to load both MCP and builtin tools</ul>

</details>

  </td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/738/files#diff-61c784391d5a45c2c6e79b1fd43056b67c8bcbf9d0558f3a8a00e8d4258517a8">+133/-23</a></td>

</tr>
</table></details></td></tr><tr><td><strong>Error handling</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>types.py</strong><dd><code>Defensive improvements to tool definition extraction</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

server/agents/base/types.py

<ul><li>Made <code>from_mcp_tool()</code> method more defensive with safer attribute access<br> <li> Added fallback to "unknown" for missing tool names using nested <br><code>getattr()</code><br> <li> Added null-coalescing for tool docstrings to handle missing <code>__doc__</code> <br>attributes</ul>

</details>

  </td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/738/files#diff-1aed621bb774ad02b47335b04dedc594edefdf271e6c380198e998e8a7d3936d">+4/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Configuration changes</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>schema.yaml</strong><dd><code>Add builtin tools configuration to model schema</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

config/schema.yaml

<ul><li>Added <code>builtin_tools</code> configuration object to Model schema<br> <li> Includes <code>enabled</code> boolean flag (default: true) as master switch for <br>builtin tools<br> <li> Includes <code>exclude</code> array for disabling specific tools by name (default: <br>empty list)</ul>

</details>

  </td>
  <td><a href="https://github.com/llama-farm/llamafarm/pull/738/files#diff-b5e001cf2a392630f0ae6d3499122dd53cb29c671ab0cff2e6e3fbc37bbdce73">+16/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tbody></table>

</details>

___

### CODE REVIEW

File: server/agents/base/types.py [hunk 1]
```diff
@@ -41,17 +41,19 @@ def from_datamodel_tool(cls, tools: Tool) -> "ToolDefinition":
...
         Args:
-            tool_class: The MCP tool class (from atomic-agents)
+            tool_class: The tool class (from atomic-agents or builtin tools)
 
         Returns:
             ToolDefinition with extracted name, description, and parameters
         """
-        tool_name = getattr(tool_class, "mcp_tool_name", tool_class.__name__)
-        tool_description = tool_class.__doc__ or "No description"
+        tool_name = getattr(
```
Comment (mhamann) [inline server/agents/base/types.py:53]: ### ⚠️ Critical: `mcp_tool_name` → `tool_name` rename may break MCP tools

`fetch_mcp_tools_async()` creates tool classes via `atomic-agents`. If the library still sets `mcp_tool_name`, this `getattr(..., "tool_name", ...)` falls back to `__name__`, silently breaking tool dispatch, `_can_execute_tool_call()`, and `_execute_mcp_tool()`.

Tests use `MagicMock` with `tool_name` set explicitly, so they **mask** this regression.

**Action:** Verify `atomic-agents` exposes `tool_name` on MCP tools. If not, use a fallback chain:
```python
tool_name = getattr(tool_class, "tool_name", None) or getattr(tool_class, "mcp_tool_name", None) or getattr(tool_class, "__name__", "unknown")
```

<details><summary>AI fix prompt</summary>

```
In server/agents/base/types.py, server/agents/chat_orchestrator.py, and server/tools/mcp_tool/tool/mcp_tool_factory.py: change all getattr(t, "tool_name", ...) lookups to try "tool_name" first, then fall back to "mcp_tool_name", then to "__name__". This ensures compatibility whether the atomic-agents library uses "tool_name" or "mcp_tool_name".
```
</details>

File: server/tools/mcp_tool/tool/mcp_tool_factory.py [hunk 2]
```diff
@@ -119,7 +119,7 @@ async def create_tools_for_server(self, server_name: str) -> list[type[BaseTool]
                 client_session=persistent_session,
             )
 
-            tool_names = [getattr(t, "mcp_tool_name", t.__name__) for t in tools]
+            tool_names = [getattr(t, "tool_name", t.__name__) for t in tools]
```
Comment (mhamann) [inline server/tools/mcp_tool/tool/mcp_tool_factory.py:122]: ### ⚠️ Critical: `voice/service.py` still uses `mcp_tool_name`

`server/api/routers/voice/service.py:656,667,676` still references `mcp_tool_name`. This file wasn't updated in this PR, creating an inconsistency — either voice or chat is broken depending on what `atomic-agents` actually exposes.

<details><summary>AI fix prompt</summary>

```
Update server/api/routers/voice/service.py to use the same attribute lookup pattern as chat_orchestrator.py for tool name resolution. Search for all occurrences of "mcp_tool_name" and replace with "tool_name" (or the fallback chain if adopted).
```
</details>

File: server/services/tasks_service.py [hunk 3]
```diff
@@ -0,0 +1,516 @@
...
+    def _read_task(cls, tasks_dir: Path, task_id: str) -> Task:
+        """
+        Read a task from disk. Must be called while holding the lock.
+
+        Raises TaskNotFoundError if the task doesn't exist.
+        """
+        task_path = tasks_dir / f"{task_id}.json"
+        if not task_path.exists():
+            raise TaskNotFoundError(f"Task '{task_id}' not found")
+        return Task(**json.loads(task_path.read_text()))
```
Comment (mhamann) [inline server/services/tasks_service.py:132]: ### 🚨 Security: Path traversal via task_id

`task_id` comes from LLM tool calls and is used directly in path construction:
```python
task_path = tasks_dir / f"{task_id}.json"
```

A value like `../../etc/foo` would read, write, or **delete** files outside the tasks directory (`delete_task` calls `task_path.unlink()`).

Fix:
```python
if not task_id.isdigit():
    raise TaskNotFoundError(f"Invalid task ID: '{task_id}'")
```

<details><summary>AI fix prompt</summary>

```
In server/services/tasks_service.py, add input validation for task_id in get_task(), update_task(), and delete_task(). Reject any task_id that is not a positive integer string using `if not task_id.isdigit(): raise TaskNotFoundError(...)`. Also validate blocker IDs in create_task() and update_task() addBlockedBy/addBlocks lists the same way.
```
</details>

File: server/services/tasks_service.py [hunk 4]
```diff
@@ -0,0 +1,516 @@
...
+        """Get the tasks directory for a session, creating it if necessary."""
+        tasks_dir = Path(project_dir) / "tasks" / session_id
+        tasks_dir.mkdir(parents=True, exist_ok=True)
+        return tasks_dir
+
+    @classmethod
+    def _get_lock(cls, tasks_dir: Path) -> FileLock:
+        """Get a file lock for the tasks directory."""
+        lock_path = tasks_dir / ".lock"
+        return FileLock(lock_path, timeout=10)
```
Comment (mhamann) [inline server/services/tasks_service.py:85]: ### High: `FileLock` is synchronous — blocks the async event loop

`FileLock.acquire()` is a blocking call. When two async coroutines hit this in the same event loop, the second blocks the **entire** loop until the first releases. Under concurrent chat sessions this causes latency spikes.

Consider wrapping lock-holding operations with `asyncio.to_thread()`, or use an `asyncio.Lock` as a first gate.

File: server/tools/builtin/tasks_tool.py [hunk 5]
```diff
@@ -0,0 +1,456 @@
...
+    # Class-level attributes for context injection
+    _project_dir: str = ""
+    _session_id: str = ""
+
+    # Tool name for identification
+    tool_name: str = "task_create"
+
+    def run(self, params: TaskCreateInput) -> TaskToolOutput:
+        """Execute the task creation synchronously."""
+        return asyncio.get_event_loop().run_until_complete(self.arun(params))
```
Comment (mhamann) [inline server/tools/builtin/tasks_tool.py]: ### High: `get_event_loop().run_until_complete()` crashes in async context

FastAPI runs in an async event loop. Calling `run_until_complete()` from within a running loop raises `RuntimeError: This event loop is already running`.

Either remove sync `run()` (if never called) or guard it. Same issue in `TaskUpdateTool`, `TaskListTool`, `TaskGetTool`.

File: config/schema.yaml [hunk 6]
```diff
@@ -270,6 +270,17 @@ properties:
                 even under memory pressure. Useful for models that need fast
                 response times (e.g., voice chat LLMs).
               default: false
+            builtin_tools:
+              type: object
+              description: |
+                Configuration for built-in tools. All builtin tools are disabled by default.
+                Use 'include' to enable specific tools by name.
```
Comment (mhamann) [inline config/schema.yaml:277]: ### High: PR description contradicts actual schema

PR description says: "`enabled: bool` master switch (default: true)" and "`exclude: list[str]`".

Actual schema uses `include: list[str]` (opt-in, all tools OFF by default). The description also says "every runtime model configuration now gets a set of built-in tools automatically" which is false.

Please update the PR description to match.

File: server/agents/chat_orchestrator.py [hunk 7]
```diff
@@ -294,8 +297,15 @@ async def run_async(
         4. Repeat until LLM provides final answer
         """
         iteration = 0
-        tools = [ToolDefinition.from_mcp_tool(t) for t in self._mcp_tools] + (
-            tools or []
+
+        # Get enabled builtin tools based on model config
+        enabled_builtin = self._get_enabled_builtin_tools()
```
Comment (mhamann) [inline server/agents/chat_orchestrator.py:302]: ### Medium: `_get_enabled_builtin_tools()` called redundantly on every tool execution

Called here to build the tool list, then called again inside `_execute_tool()` and `_can_execute_tool_call()` on **every** tool call within the same request. Compute once and pass through, or cache per-request.

File: server/tools/builtin/factory.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+        if self._session_id is None:
+            return None
+
+        project_dir = self._project_dir
+        session_id = self._session_id
+
+        class InjectedTool(base_class):  # type: ignore[valid-type]
+            _project_dir = project_dir
+            _session_id = session_id
+
```
Comment (mhamann) [inline server/tools/builtin/factory.py:52]: ### Medium: Injected tool classes lose their `__name__`

All injected tools get `__name__ = "InjectedTool"`. Add after class creation:
```python
InjectedTool.__name__ = base_class.__name__
InjectedTool.__qualname__ = base_class.__qualname__
```

File: server/project.json [hunk 9]
```diff
@@ -23,11 +23,18 @@
...
+        "cwd": "{workspaceRoot}/server",
+        "command": "uv run ruff check ."
+      }
+    },
     "test": {
       "executor": "nx:run-commands",
       "options": {
         "cwd": "{workspaceRoot}/server",
-        "command": "uv run pytest -v --cov=. --cov-report=term-missing"
+        "command": "uv run pytest -v"
```
Comment (mhamann) [inline server/project.json:37]: ### Medium: Coverage reporting silently removed

`--cov=. --cov-report=term-missing` was dropped. Intentional? If so, call it out in the PR description.

File: server/tests/test_chat_orchestrator_agent.py [hunk 10]
```diff
@@ -533,6 +554,100 @@ async def second_stream(*args, **kwargs):
...
+                    chunks.append(chunk)
+
+                # The tool should have been called with the complete arguments
+                mock_tool_instance.arun.assert_awaited()
+
+                # Verify the tool was called with the correct accumulated arguments
+                call_args = mock_tool_instance.arun.call_args
+                assert call_args is not None
+                # The tool receives the parsed arguments
+                assert "operation" in str(call_args) or mock_tool_instance.arun.called
```
Comment (mhamann) [inline server/tests/test_chat_orchestrator_agent.py]: ### Low: Assertion is always true

`mock_tool_instance.arun.called` is already asserted on line above, so the `or` makes this vacuously true. Assert just the specific thing:
```python
assert "operation" in str(call_args)
```

File: server/services/tasks_service.py [hunk 11]
```diff
@@ -0,0 +1,516 @@
...
+        task_path = tasks_dir / f"{task.id}.json"
+        task_path.write_text(json.dumps(task.model_dump(), indent=2))
+
+    @classmethod
+    def _validate_no_cycles(
+        cls,
+        tasks_dir: Path,
+        task_id: str,
+        new_blocked_by: list[str],
+    ) -> None:
```
Comment (mhamann) [inline server/services/tasks_service.py:146]: ### Medium: Cycle detection reads all task files on every call

Deserializes every `*.json` in the directory each time. Gets expensive with many tasks. The recursive DFS also has no depth limit — a long dependency chain could hit Python's recursion limit (default 1000).

File: server/project.json (hunk 9 above)
Comment (BobbyRadford) [reply server/project.json:37]: intentional

**Review by mhamann — APPROVED** [review]
